WIP: online-cmvn for online2 with chain models,

vesis84 opened the pull request:
- successor of #3439 from @luitjens , 
- this PR includes udpate of 'prepare_online_decoding.sh'
- it is not not yet fully tested (currently running)
- @danpovey Please take a look at the code, i'll do changes on request...

danpovey commented:
Karel, can you please take a look at issue #3554 and PR #3462, figure out what issues remain in the light of that, and let me know what the key choices/decisions are?  I.e. is your approach different from #3462? 

@freewym was having trouble getting #3462 to work right (in a setup of his without ivectors) but I don't know if he got the latest fix.  @freewym if you have time, please also have a look and let me know whether Karel has made the same choices as you+Levi.

Also of course we need to take into account the speaker information when applying online CMVN, and use the _max2 features in the scripts, as I pointed out in #3554.

vesis84 commented:
This PR refactors the original code from #3462, and it does not use spk2utt for porting cmvn stats (suggested in #3554). Can we accept this PR and add porting of cmvn stats in a separate PR?

vesis84 commented:
Hmm, there seems to be a bug, the WER is high (~70%)... I need to investigate...

danpovey commented:
Let's see what @freewym says first...
yes, we could make the speaker info a separate issue.


On Tue, Sep 3, 2019 at 4:55 PM Karel Vesely <notifications@github.com>
wrote:

> Hmm, there seems to be a bug, the WER is high (~70%)... I need to
> investigate...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/3560?email_source=notifications&email_token=AAZFLO2NTWM76AMN4DIGB6LQHYRAVA5CNFSM4IS6R2J2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD5XQGKY#issuecomment-527369003>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAZFLO6BXSLVCUU3AQDVHKDQHYRAVANCNFSM4IS6R2JQ>
> .
>

freewym commented on src/online2/online-nnet2-feature-pipeline.cc:
```diff
@@ -149,15 +174,28 @@ void OnlineNnet2FeaturePipeline::GetAdaptationState(
   // Some of the online-feature pointers are just copies of other pointers,
   // and we do have to avoid deleting them in those cases.
-  if (final_feature_ != feature_plus_optional_pitch_)
+  if (final_feature_ != feature_plus_optional_cmvn_)
```
Probably some of the memory release code here should handle the NULL pointer case, like [here](https://github.com/kaldi-asr/kaldi/pull/3462/files#diff-297989e9a4e929019e7b6db4e1b549f4R214), e.g., `ivector_feature_`, `feature_plus_optional_cmvn_`, `pitch_feature_`, `pitch_`, etc

freewym commented on src/online2/online-nnet2-feature-pipeline.h:
```diff
@@ -265,19 +283,33 @@ class OnlineNnet2FeaturePipeline: public OnlineFeatureInterface {
+  /// Init() is to be called from the constructor; it assumes the pointer
+  /// members are all uninitialized but config_ and lda_mat_ are
+  /// initialized.
+  void Init();
```
This function was never used or implemented.

vesis84 replied:
Thanks! Now it makes more sense, why it did not work :P

Anyway, why is there the `final_feature_` if it is not used by the nnet3 codebase? 
Is it to enforce same number of frames in both feature streams?
K.

freewym commented on src/online2/online-nnet2-feature-pipeline.cc:
```diff
@@ -76,9 +84,14 @@ OnlineNnet2FeaturePipelineInfo::OnlineNnet2FeaturePipelineInfo(
+    pitch_(NULL), pitch_feature_(NULL),
+    cmvn_feature_(NULL), ivector_feature_(NULL) {
+
   if (info_.feature_type == "mfcc") {
```
maybe move most of these code to Init(), to keep consistency of its parent class OnlineFeaturePipeline 
 in src/online2/online-feature-pipeline.cc

danpovey commented:
Thanks!
BTW, in C++, `delete` is just a no-op for NULL argument.

On Wed, Sep 4, 2019 at 10:52 AM Yiming Wang <notifications@github.com>
wrote:

> *@freewym* commented on this pull request.
> ------------------------------
>
> In src/online2/online-nnet2-feature-pipeline.cc
> <https://github.com/kaldi-asr/kaldi/pull/3560#discussion_r320552925>:
>
> >  OnlineNnet2FeaturePipeline::OnlineNnet2FeaturePipeline(
>      const OnlineNnet2FeaturePipelineInfo &info):
> -    info_(info) {
> +    info_(info), base_feature_(NULL),
> +    pitch_(NULL), pitch_feature_(NULL),
> +    cmvn_feature_(NULL), ivector_feature_(NULL) {
> +
>    if (info_.feature_type == "mfcc") {
>
> maybe move most of these code to Init(), to keep consistency of its parent
> class OnlineFeaturePipeline
> in src/online2/online-feature-pipeline.cc
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/3560?email_source=notifications&email_token=AAZFLO5A6EBBC2SDHC4CCELQH4PFJA5CNFSM4IS6R2J2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCDRZ6SQ#pullrequestreview-283352906>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAZFLO34THFXM4ALLJCKXKLQH4PFJANCNFSM4IS6R2JQ>
> .
>

vesis84 commented:
Now it makes more sense, why it did not work :P
I assumed that only `final_feature_` is accessed 
by then nnet3 classes, which a false assumption...

Thanks!
K.

vesis84 commented:
Okay, so I removed the `Init()` and added the `nnet3_feature_` pointer, to prevent the same problem happen again... (make it more obvious which stream goes on nnet3 input).

The WER tests are good, but the adaptation stats are not transferred between utts of same speaker.
@danpovey Shall we accept the PR code as it is now? And solve the transfer in another PR?

vesis84 commented:
This improved the results:
10.96 -> 10.60 tgsmall
7.44 -> 7.22 tglarge
(the models were trained without the transfer of the OnlineCmvnState)

vesis84 commented:
Now running the training which includes the transfer of 'OnlineCmvnState'. Results will be tomorrow...
K.

freewym commented on egs/wsj/s5/steps/nnet3/chain/get_egs.sh:
```diff
@@ -221,9 +221,9 @@ else
-  feats="ark,s,cs:utils/filter_scp.pl --exclude $dir/valid_uttlist $sdata/JOB/feats.scp | apply-cmvn-online $cmvn_opts $dir/global_cmvn.stats scp:- ark:- |"
-  valid_feats="ark,s,cs:utils/filter_scp.pl $dir/valid_uttlist $data/feats.scp | apply-cmvn-online $cmvn_opts $dir/global_cmvn.stats scp:- ark:- |"
-  train_subset_feats="ark,s,cs:utils/filter_scp.pl $dir/train_subset_uttlist $data/feats.scp | apply-cmvn-online $cmvn_opts $dir/global_cmvn.stats scp:- ark:- |"
+  feats="ark,s,cs:utils/filter_scp.pl --exclude $dir/valid_uttlist $sdata/JOB/feats.scp | apply-cmvn-online $cmvn_opts --spk2utt=ark:$sdata/JOB/spk2utt $dir/global_cmvn.stats scp:- ark:- |"
```
Please also apply similar changes to nnet3/chain/e2e/get_egs_e2e.sh

vesis84 replied:
okay, now it is added (i also removed the option `$feat_type`, it was unused)

danpovey commented:
Great!
I think I'm OK to merge this as long as in test time for online decoding it is carrying forward the CMVN state... just so we have one fewer thing to change in future (i.e. not to check in code that we know will change).
Incidentally, @freewym tested online-CMVN on AMI and it got slightly worse.. probably because it's recognizing background speech by other speakers.

danpovey commented:
... oh, and I'll wait till @freewym says it's OK to merge.  I'm too busy right now to go through it myself (I'm in Beijing right now.)

vesis84 commented:
> This improved the results:
> 10.96 -> 10.60 tgsmall
> 7.44 -> 7.22 tglarge
> (the models were trained without the transfer of the OnlineCmvnState)

Adding the transfer of OnlineCmvnState into the training:
10.60 -> 10.62 tgsmall
7.20 -> 7.16 tglarge
=> no change! (but let's keep it in, so that training and online2 systems handle OnlineCmvnState in same way)

vesis84 commented:
From my point of view, it is done. Please let me know if you need to change something...
Best, Karel

vesis84 commented:
But still, I could re-run the recipes, and update the results that will be better now...

danpovey commented:
Thanks!
I'll wait till I hear from Yiming about what he thinks I should do.
I know he did much of this himself already, in parallel, but if he thingk
yours can be merged, that's fine with me.


On Thu, Sep 5, 2019 at 6:42 PM Karel Vesely <notifications@github.com>
wrote:

> But still, I could re-run the recipes, and update the results that will be
> better now...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/3560?email_source=notifications&email_token=AAZFLO5KMCNCOXKI4DEQ63DQIDPA7A5CNFSM4IS6R2J2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD56VBUA#issuecomment-528306384>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAZFLO3RVX7XJX7GAZ3GHNLQIDPA7ANCNFSM4IS6R2JQ>
> .
>

freewym commented on src/online2/online-nnet2-feature-pipeline.h:
```diff
@@ -241,58 +263,75 @@ class OnlineNnet2FeaturePipeline: public OnlineFeatureInterface {
+  /// The following constructor is used internally in the New() function;
+  /// it has the same effect as initializing from just "cfg", but avoids
+  /// re-reading the LDA transform from disk.
+  OnlineNnet2FeaturePipeline(const OnlineNnet2FeaturePipelineInfo &info_,
```
This declaration can be removed as well, as it has never be used.

freewym commented:
> Thanks! I'll wait till I hear from Yiming about what he thinks I should do. I know he did much of this himself already, in parallel, but if he thingk yours can be merged, that's fine with me.
> […](#)
> On Thu, Sep 5, 2019 at 6:42 PM Karel Vesely ***@***.***> wrote: But still, I could re-run the recipes, and update the results that will be better now... — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <#3560?email_source=notifications&email_token=AAZFLO5KMCNCOXKI4DEQ63DQIDPA7A5CNFSM4IS6R2J2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD56VBUA#issuecomment-528306384>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAZFLO3RVX7XJX7GAZ3GHNLQIDPA7ANCNFSM4IS6R2JQ> .

I am fine with this PR to be merged. Just added a little bit more comments.

danpovey commented:
@freewym is there any chance you could run something with this PR?  E.g. one of your kws experiments?  Just so we have a little more certainty that it's all debugged.

freewym commented:
Ok. Will test with speaker info carrying over utts. But the wake word
recipe doesn’t use ivectors.

On Thu, Sep 5, 2019 at 8:18 PM Daniel Povey <notifications@github.com>
wrote:

> @freewym <https://github.com/freewym> is there any chance you could run
> something with this PR? E.g. one of your kws experiments? Just so we have a
> little more certainty that it's all debugged.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/3560?email_source=notifications&email_token=AA2YBEXZYABTNTW6BCGNTNDQIGOVXA5CNFSM4IS6R2J2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6BKN7A#issuecomment-528656124>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA2YBETQXWNA7GSDVNO7L4DQIGOVXANCNFSM4IS6R2JQ>
> .
>
-- 
Sent from my iPhone

LeviBarnes commented on src/online2/online-nnet2-feature-pipeline.cc:
```diff
@@ -64,17 +64,15 @@ OnlineNnet2FeaturePipelineInfo::OnlineNnet2FeaturePipelineInfo(
-  } else { // else user the defaults
-    use_cmvn = false;
+      KALDI_ERR << "--global-cmvn-stats option is required "
+                << " when --cmvn-config is specified.";
```
@danpovey suggested we change this to --feat-cmvn-config

LeviBarnes commented on src/online2/online-nnet2-feature-pipeline.h:
```diff
@@ -104,10 +108,12 @@ struct OnlineNnet2FeaturePipelineConfig {
     opts->Register("fbank-config", &fbank_config, "Configuration file for "
                    "filterbank features (e.g. conf/fbank.conf)");
     opts->Register("cmvn-config", &cmvn_config, "Configuration file for "
```
At the suggestions of @danpovey, we changed this to feat-cmvn-config

danpovey replied:
Hm.  If this setup is already tested, I might be OK to have the name stay as it is.  I was not 100% sure/confident about that change actualy.

vesis84 replied:
Yes, I would prefer to keep it as it is too (CMVN is always applied to features, so the word 'feat' does not bring more information)

LeviBarnes replied:
Great by me

vesis84 commented:
I have the updated results.
```
orig  no-transfer with-transfer
10.97 10.75       10.64     mini_librispeech, dev_clean_2, tgsmall
 8.32  8.22        8.50    tedlium, dev(orig)
 8.44  8.45        8.39    tedlium, test(orig)
 6.89  6.66        6.90    wsj, dev93 (tgpr)
 4.68  4.89        4.77    wsj, eval92 (tgpr)
```
orig = no online-cmvn, just i-vector adaptation,
no-transfer = online-cmn, done per utterance smoothed by global stats,
with-transfer = online-cmn, with accumulating online-cmvn stats over all data of a speaker,

It is interesting that adding the 'transfer' of OnlineCmvnState did not always help.
And it is not clear what is the best.

Also, I've noticed that in `steps/online/nnet2/train_ivector_extractor.sh` the original feature pipeline does not transfer the OnlineCmvnState (no --spk2utt specified), even when the nnet3 model transfers it.

Should I activate the 'transfer' here as well?
K.

freewym commented:
I tested it with wake word experiment and it works.

danpovey commented:
Thanks, guys!  I'll merge...
we can test the fix to the ivector-extractor training script separately.
It probably won't make much difference since when the ivectors were
extracted it did carry forward that info, but it is a mismatch that should
be fixed.

On Sat, Sep 7, 2019 at 7:18 AM Levi Barnes <notifications@github.com> wrote:

> *@LeviBarnes* commented on this pull request.
> ------------------------------
>
> In src/online2/online-nnet2-feature-pipeline.h
> <https://github.com/kaldi-asr/kaldi/pull/3560#discussion_r321938622>:
>
> > @@ -104,10 +108,12 @@ struct OnlineNnet2FeaturePipelineConfig {
>      opts->Register("fbank-config", &fbank_config, "Configuration file for "
>                     "filterbank features (e.g. conf/fbank.conf)");
>      opts->Register("cmvn-config", &cmvn_config, "Configuration file for "
>
> Great by me
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/3560?email_source=notifications&email_token=AAZFLO465IBFB4HEDEBPOT3QILQL7A5CNFSM4IS6R2J2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCD7NJAA#discussion_r321938622>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAZFLO2HTZEII7464Y2HPBLQILQL7ANCNFSM4IS6R2JQ>
> .
>
